src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.cs

```diff
@@ -78,33 +77,18 @@ private async Task ProcessAsync(ProjectToDocumentMap projectToDocumentMap)
                     return;
                 }
 
-                // Get the connected components of the dependency graph and process each individually.
-                // That way once a component is done we can throw away all the memory associated with
-                // it.
-                // For each connected component, we'll process the individual projects from bottom to
-                // top.  i.e. we'll first process the projects with no dependencies.  Then the projects
-                // that depend on those projects, and so on.  This way we always have created the 
-                // dependent compilations when they're needed by later projects.  If we went the other
-                // way (i.e. processed the projects with lots of project dependencies first), then we'd
-                // have to create all their dependent compilations in order to get their compilation.
-                // This would be very expensive and would take a lot of time before we got our first
-                // result.
-                var connectedProjects = _dependencyGraph.GetDependencySets(_cancellationToken);
-
                 // Add a progress item for each (document, symbol, finder) set that we will execute.
                 // We'll mark the item as completed in "ProcessDocumentAsync".
                 var totalFindCount = projectToDocumentMap.Sum(
                     kvp1 => kvp1.Value.Sum(kvp2 => kvp2.Value.Count));
                 await _progressTracker.AddItemsAsync(totalFindCount).ConfigureAwait(false);
 
-                // Now, go through each connected project set and process it independently.
-                foreach (var connectedProjectSet in connectedProjects)
-                {
-                    _cancellationToken.ThrowIfCancellationRequested();
+                using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
 
-                    await ProcessProjectsAsync(
-                        connectedProjectSet, projectToDocumentMap).ConfigureAwait(false);
-                }
+                foreach (var (project, documentMap) in projectToDocumentMap)
+                    tasks.Add(Task.Run(() => ProcessProjectAsync(project, documentMap), _cancellationToken));
```

Comment:
Is Task.Run here just to pass cancellationToken?

Reply:
It's to ensure that the work is truly parallel and the main lol can get to the next item to add to the work queue. If there work doesn't actually ever yield, and if entirely synchronous, then not doing this could have no benefit.

Reply:
>It's to ensure that the work is truly parallel and the main lol can get to the next item to add to the work queue. 

I looked back here to read your answer again. But I think that's not what I was asking.

I was wondering why not: `tasks.Add(ProcesssProjectAsync(...))`? Is cancellationToken the only reason?

Reply:
> I was wondering why not: tasks.Add(ProcesssProjectAsync(...))? Is cancellationToken the only reason?
@CyrusNajmabadi


No.  It's so the work is *actually* parallel.  Adding lots of async work doesn't make anything acutally parallel.  Consider something as simple as this:

```
public Task ExpensiveAsync()
{
    Thread.Sleep(10000);
    return Task.CompletedTask;
}
```

if we go in a loop 10 times and attempt to add ExpensiveAsync to an array of tasks, it will take 100 seconds.  By doing Task.Run, we explicitly move this work to a threadpool thread, and allow the work to operate entirely concurrently, allowing it all to finish in 10 seconds.

This is obviously an extreme case (total thread blocking, no actual asynchrony), but the general idea still holds.  Putting a lot of async work into a queue doesn't make it necessary concurrent.  Doing the Task.Run does ensure this.

Reply:
>This is obviously an extreme case (total thread blocking, no actual asynchrony), but the general idea still holds.

Got it. Just so that I understand, it does depend on the ProcesssProjectAsync implementation to be possibly CPU-bound/Blocking, and even for IO-bound scenarios, Task.Run will ensure any "additional work" that is being done as part of it to run in parallel.

Reply:
Yes.  `tasks.Add(ProcesssProjectAsync(...))` is possibly concurrent.  `tasks.Add(Task.Run(() ` is definitely concurrent.